test/replication-luatest/join_test.lua

```diff
@@ -107,3 +107,38 @@ g.test_raft_state_during_meta_join = function(lg)
     replica:drop()
     lg.master:update_box_cfg{election_mode = 'off'}
 end
+
+-- The test covers a crash on two consequents JOIN requests with
+-- the same UUID.
+g.test_join_with_the_same_uuid = function(lg)
+    t.tarantool.skip_if_not_debug()
+    lg.master:exec(function()
+        local s = box.schema.create_space('test')
+        s:create_index('pk')
+        s:replace{1}
+        box.error.injection.set('ERRINJ_REPLICA_JOIN_DELAY', true)
+    end)
+    local replica_uuid = uuid.str()
+    local box_cfg = table.deepcopy(lg.master.box_cfg)
+    box_cfg.replication = {lg.master.net_box_uri}
+    box_cfg.instance_uuid = replica_uuid
+    local replica1 = server:new({
+        alias = 'replica1',
+        box_cfg = box_cfg,
+    })
+    local replica2 = server:new({
+        alias = 'replica2',
+        box_cfg = box_cfg,
+    })
+    replica1:start({wait_until_ready = false})
+    replica2:start({wait_until_ready = false})
+    local msg = ('joining replica'):format(replica_uuid):gsub('%-', '%%-')
+    t.helpers.retrying({}, function()
+        t.assert(lg.master:grep_log(msg))
```

Comment:
As you can see, I not only introduce a new case here but also turn `before_all` hook into `before_each`. The reason of the failure you sent is that the cases are not isolated so old cluster topology breaks it.

> Moreover, grep_log doesn't checks, that there's exactly one expected log message in logs

I don't know how to check if a string appears in logs exactly once.

> Isn't it a duplicate of `g.test_fetch_snapshot_on_connected_replica`.

The idea is the same, but this case covers IPROTO_JOIN.

> We don't assert, that error on one of replicas happened, just that master starts to join replica.

The second replica infinitely tries to join but it can't because another replica with the same UUID is connected. Until the join succeeds, any attempt to do `replica2:exec` fails with "net box not connected" error.